engine/src/scriptrunner.h

```diff
@@ -199,24 +224,45 @@ public slots:
      */
     int random(int minTime, int maxTime);
 
+protected slots:
+    /** Triggered when the script's execution pauses to await the starting of a function */
+    void slotWaitFunctionStarted(quint32 fid);
+
+    /** Triggered when the script's execution pauses to await the completion of a function */
+    void slotWaitFunctionStopped(quint32 fid);
+
 protected:
     /** QThread reimplemented method */
     void run();
 
 private:
+    /** ScriptRunner function operations enum to handle start/stop/wait commands */
+    enum class SRFuncOpe : char
+    {
+        START,
+        START_DONT_STOP,
+        STOP,
+        WAIT_START,
+        WAIT_STOP
+    };
+
     Doc *m_doc;
     QString m_content;
     bool m_running;
 
     QJSEngine *m_engine;
     // Queue holding the Function IDs to start/stop
-    QQueue<QPair<quint32,bool>> m_functionQueue;
+    QQueue<QPair<quint32, SRFuncOpe>> m_functionQueue;
     // Queue holding Fixture values to send to Universes
     QQueue<FixtureValue> m_fixtureValueQueue;
+    // Indicate to add (true) or to not add (false) to the Functions started by this script
+    bool m_stopOnExit;
     // IDs of the Functions started by this script
     QList <quint32> m_startedFunctions;
     // Timer ticks to wait before executing the next line
     quint32 m_waitCount;
+    // Function that the script is waiting for
+    Function* m_waitFunction;
```

Comment:
I don't like this. Instead of saving pointer to Function, just save the quint32 functionID. (and remove the include to function.h)
Then instead of using Function signals, use these instead:
https://github.com/mcallegari/qlcplus/blob/master/engine/src/mastertimer.h#L115
https://github.com/mcallegari/qlcplus/blob/master/engine/src/mastertimer.h#L118

And compare IDs rather than working on pointers

Reply:
Fixed on commit f3d3ca0d0fae741279f2ad6434bf0eed902cd16b